extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,221 @@
+import {
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './stream-gallery.jss';
+
+const DEFAULT_MEASUREMENT = {
+  visibleCount: 1,
+  maxContainerWidth: Number.MAX_VALUE,
+};
+
+/**
+ * @param {!StreamGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function StreamGallery({
+  arrowPrev: customArrowPrev,
+  arrowNext: customArrowNext,
+  children,
+  extraSpace,
+  insetArrowVisibility,
+  loop,
+  maxItemWidth = Number.MAX_VALUE,
+  minItemWidth = 1,
+  maxVisibleCount = Number.MAX_VALUE,
+  minVisibleCount = 1,
+  outsetArrows,
+  peek = 0,
+  snap,
+  style,
+  ...rest
+}) {
+  const galleryRef = useRef(null);
+  const galleryWidthRef = useRef(0);
+  const [measurements, setMeasurements] = useState(DEFAULT_MEASUREMENT);
+  const arrowPrev = useMemo(
+    () =>
+      customArrowPrev || <DefaultArrow by={-1} outsetArrows={outsetArrows} />,
+    [customArrowPrev, outsetArrows]
+  );
+  const arrowNext = useMemo(
+    () =>
+      customArrowNext || <DefaultArrow by={1} outsetArrows={outsetArrows} />,
+    [customArrowNext, outsetArrows]
+  );
+
+  const measure = useCallback(
+    () =>
+      getVisibleCount(
+        maxItemWidth,
+        minItemWidth,
+        maxVisibleCount,
+        minVisibleCount,
+        children.length,
+        peek,
+        galleryWidthRef.current
+      ),
+    [
+      maxItemWidth,
+      minItemWidth,
+      maxVisibleCount,
+      minVisibleCount,
+      children.length,
+      peek,
+    ]
+  );
+
+  // Adjust visible slide count when container size or parameters change.
+  useLayoutEffect(() => {
+    const node = galleryRef.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver(() => {
+      if (galleryWidthRef.current === node./* OK */ offsetWidth) {
+        return;
+      }
+      galleryWidthRef.current = node./* OK */ offsetWidth;
+      setMeasurements(measure());
+    });
+    observer.observe(node);
+    return () => observer.disconnect();
+  }, [measure]);
+
+  return (
+    <div ref={galleryRef}>
+      <BaseCarousel
+        advanceCount={Math.floor(measurements.visibleCount)}
+        arrowPrev={arrowPrev}
+        arrowNext={arrowNext}
+        controls={insetArrowVisibility}
+        loop={loop}
+        outsetArrows={outsetArrows}
+        snap={snap}
+        style={{
+          ...style,
+          flexGrow: 1,
+          maxWidth:
+            measurements.maxContainerWidth >= Number.MAX_VALUE
+              ? ''
+              : measurements.maxContainerWidth,
+          justifyContent: extraSpace === 'around' ? 'center' : 'initial',
+        }}
+        visibleCount={measurements.visibleCount}
+        {...rest}
+      >
+        {children}
+      </BaseCarousel>
+    </div>
+  );
+}
+
+/**
+ * @param {!StreamGalleryDef.ArrowProps} props
+ * @return {PreactDef.Renderable}
+ */
+function DefaultArrow({advance, by, outsetArrows, ...rest}) {
```

Comment:
Is this different from the carousel's?

Reply:
Yes, it is inverted - black arrow on white button versus white arrow on black button. It also has no additional `<div>` layers, so we should update the tracking issue to see if those can be removed in `BaseCarousel` as well without much change in appearance.

Reply:
Also the inset ones are square instead of circular:
![image](https://user-images.githubusercontent.com/10456171/97187660-0b1f1480-1779-11eb-8232-7439873f8c20.png)

Outset:
![image](https://user-images.githubusercontent.com/10456171/97187668-0d816e80-1779-11eb-86d0-f204d331ab91.png)